x/ratelimit/ics4_wrapper.go

```diff
@@ -0,0 +1,38 @@
+package ratelimit
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	porttypes "github.com/cosmos/ibc-go/v3/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+
+	ratelimitkeeper "github.com/Stride-Labs/stride/v4/x/ratelimit/keeper"
+)
+
+var (
+	_ porttypes.Middleware  = &IBCModule{}
+	_ porttypes.ICS4Wrapper = &ICS4Wrapper{}
+)
+
+type ICS4Wrapper struct {
+	channel         porttypes.ICS4Wrapper
+	rateLimitKeeper ratelimitkeeper.Keeper
+}
+
+func NewICS4Middleware(channel porttypes.ICS4Wrapper, ratelimitKeeper ratelimitkeeper.Keeper) ICS4Wrapper {
+	return ICS4Wrapper{
+		channel:         channel,
+		rateLimitKeeper: ratelimitKeeper,
+	}
+}
+
+func (i *ICS4Wrapper) SendPacket(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet exported.PacketI) error {
```

Comment:
Why is this `SendPacket` used, rather than the `SendPacket` in `ibc_module.go`?

Reply:
This was by far the most confusing aspect of the ICS Middleware setup. It has to do with the ordering of the execution of each stack: sends are bottom-up, receives are top-down.

**Receive Packet**
1. IBC Channel Module
2. OnRecvPacket in ratelimit module
3. OnRecvPacket in records module
4. OnRecvPacket in transfer module
5. Tokens are minted

**Send Packet**
1. SendPacket from transfer module
2. Tokens are escrowed
3. SendPacket from ICS4 Wrapper (rate limit module)
4. SendPacket from Channel Module

---

In the receive scenario, the execution flow aligns with the chaining defined when setting up and wrapping each middleware. So the **rate limit** module has access to the **records** module, which has access to the **transfer** module.
```go
stack := ratelimit.NewIBCModule(rateLimitKeeper, records.NewIBCModule(recordsKeeper, transferModule))
```

But in the send scenario, the transfer module doesn't have access to that full middleware chaining (since it's at the bottom of the stack) - it only has access to its ICSMiddleware.
```go
transferKeeper := NewKeeper(..., icsMiddleware, ...)
```

That's my understanding of it at least - I'm still not convinced I 100% understand it haha